File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        seen = set()
        ROWS, COLS = len(grid), len(grid[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    
        def dfs(r, c):
            if (r < 0 or r >= ROWS or
                c < 0 or c >= COLS or
                grid[r][c] == 0 or 
                (r, c) in seen):
                return 0
            seen.add((r, c))
            area = 1
            for dr, dc in directions:
                row, col = r + dr, c + dc
                area += dfs(row, col)


            return area





        maxArea = 0
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) not in seen and grid[r][c] == 1:
                    maxArea = max(maxArea, dfs(r, c))
        return maxArea
    
    
        

File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    int ROWS;
    int COLS;
    vector<vector<int>> grid;
    vector<vector<int>> dir = {{1, 0},{-1, 0},{0, 1},{0, -1}};
    
    int dfs(int r, int c, vector<vector<bool>> &visited) {
        visited[r][c] = true;
        int ans = 1;
        for (auto d: dir) {
            int row = r + d[0];
            int col = c + d[1];
            if (row >= 0 && row < ROWS && col >= 0 && col < COLS && !visited[row][col] && grid[row][col] == 1) {
                ans += dfs(row, col, visited);
            }
        }
        return ans;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        this->grid = grid;
        ROWS = grid.size();
        COLS = grid[0].size();
        vector<vector<bool>> visited(ROWS, vector<bool>(COLS, false));
        int res = 0;
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (!visited[r][c] && grid[r][c] == 1) {
                    res = max(res, dfs(r, c, visited));
                }
            }
        }
        return res;
    }
};